src/eval/operation.rs

```diff
@@ -1119,14 +1124,29 @@ impl<R: ImportResolver, C: Cache> VirtualMachine<R, C> {
                         Term::Record(record) if !record.fields.is_empty() => {
                             let mut shared_env = Environment::new();
 
-                            let record = record.map_values_closurize(&mut self.cache, &mut shared_env, &env, |_, t| {
-                                mk_term::op1(UnaryOp::Force(), t)
-                            }).map_err(|missing_field_err| missing_field_err.into_eval_err(pos, pos_op))?;
+                            let fields = record.fields
+                                .into_iter()
+                                .filter(|(_, field)| {
+                                    !(field.is_empty_optional() || (for_export && field.metadata.not_exported))
+                                })
+                                .map_values_closurize(&mut self.cache, &mut shared_env, &env, |_, value| {
+                                    mk_term::op1(UnaryOp::Force { for_export }, value)
+                                })
+                                .map_err(|e| e.into_eval_err(pos, pos_op))?;
+
+                            let terms = fields
+                                .clone()
+                                .into_values()
+                                .map(|field| {
+                                    field
+                                        .value
+                                        .expect("fields without definition would have thrown MissingFieldDefError before")
+                                });
 
-                            // unwrap: the call to map_fields_without_optionals must ensure that
```

Comment:
Nitpick: I think we lose a bit of information here by removing the comment. The expect message doesn't talk about `map_fields_without_optionals` at all, for example. Although I guess it's now called `map_values_closurize` ? Anyway, I would suggest to either keep both (comment + message), or to extend the message of `expect` to provide at least as much context as before.

Reply:
I've changed the `expect` message to mention that `map_values_closurize` enforces the invariant. The comment had indeed gone stale, `map_fields_without_optionals` got refactored to `map_values_closurize` at some point.

Reply:
Sounds good